contracts/helpers/midas/MidasRedemptionVaultGateway.sol

```diff
@@ -0,0 +1,226 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Gearbox Protocol. Generalized leverage for DeFi protocols
+// (c) Gearbox Foundation, 2024.
+pragma solidity ^0.8.23;
+
+import {ReentrancyGuardTrait} from "@gearbox-protocol/core-v3/contracts/traits/ReentrancyGuardTrait.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {IERC20Metadata} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IMidasRedemptionVault} from "../../integrations/midas/IMidasRedemptionVault.sol";
+import {IMidasRedemptionVaultGateway} from "../../interfaces/midas/IMidasRedemptionVaultGateway.sol";
+
+import {WAD} from "@gearbox-protocol/core-v3/contracts/libraries/Constants.sol";
+
+/// @title Midas Redemption Vault Gateway
+/// @notice Gateway contract that manages redemptions from Midas vault on behalf of other accounts
+/// @dev Stores pending redemption requests and handles partial withdrawals
+contract MidasRedemptionVaultGateway is ReentrancyGuardTrait, IMidasRedemptionVaultGateway {
+    using SafeERC20 for IERC20;
+
+    bytes32 public constant override contractType = "GATEWAY::MIDAS_REDEMPTION_VAULT";
+    uint256 public constant override version = 3_10;
+
+    address public immutable midasRedemptionVault;
+    address public immutable mToken;
+
+    mapping(address => PendingRedemption) public pendingRedemptions;
+
+    /// @notice Constructor
+    /// @param _midasRedemptionVault Address of the Midas Redemption Vault
+    constructor(address _midasRedemptionVault) {
+        midasRedemptionVault = _midasRedemptionVault;
+        mToken = IMidasRedemptionVault(_midasRedemptionVault).mToken();
+    }
+
+    /// @notice Performs instant redemption of mToken for output token
+    /// @param tokenOut Output token to receive
+    /// @param amountMTokenIn Amount of mToken to redeem
+    /// @param minReceiveAmount Minimum amount of output token to receive
+    /// @dev Transfers mToken from sender, redeems, and transfers output token back
+    function redeemInstant(address tokenOut, uint256 amountMTokenIn, uint256 minReceiveAmount) external nonReentrant {
+        IERC20(mToken).safeTransferFrom(msg.sender, address(this), amountMTokenIn);
+
+        uint256 balanceBefore = IERC20(tokenOut).balanceOf(address(this));
+
+        IERC20(mToken).forceApprove(midasRedemptionVault, amountMTokenIn);
```

Comment:
At least based on mHYPER's implementation, it seems we can give the vault a `type(uint256).max` allowance in constructor and not do it in every (instant or delayed) request.